### ISSUE TITLE
Errors when compiling with CUDA 11.6

### ISSUE THREAD

**Issue by erikguo** [triton-inference-server/paddlepaddle_backend#5]
When we build paddlepaddle v2.2.2 with CUDA 11.6, it always has errors when we turn on TENSORRT. Could you give us some suggestions to resolve it ? Thank you very much!

```
In file included from /usr/local/cuda/targets/x86_64-linux/include/thrust/system/cuda/detail/util.h:36,
                 from /usr/local/cuda/targets/x86_64-linux/include/thrust/system/cuda/detail/internal/copy_cross_system.h:41,
                 from /usr/local/cuda/targets/x86_64-linux/include/thrust/system/cuda/detail/copy.h:100,
                 from /usr/local/cuda/targets/x86_64-linux/include/thrust/system/detail/adl/copy.h:42,
                 from /usr/local/cuda/targets/x86_64-linux/include/thrust/detail/copy.inl:22,
                 from /usr/local/cuda/targets/x86_64-linux/include/thrust/detail/copy.h:90,
                 from /usr/local/cuda/targets/x86_64-linux/include/thrust/detail/allocator/copy_construct_range.inl:21,
                 from /usr/local/cuda/targets/x86_64-linux/include/thrust/detail/allocator/copy_construct_range.h:45,
                 from /usr/local/cuda/targets/x86_64-linux/include/thrust/detail/contiguous_storage.inl:23,
                 from /usr/local/cuda/targets/x86_64-linux/include/thrust/detail/contiguous_storage.h:234,
                 from /usr/local/cuda/targets/x86_64-linux/include/thrust/detail/vector_base.h:30,
                 from /usr/local/cuda/targets/x86_64-linux/include/thrust/device_vector.h:26,
                 from /workspace/Paddle/paddle/fluid/inference/tensorrt/plugin/gather_nd_op_plugin.h:17,
                 from /workspace/Paddle/paddle/fluid/inference/tensorrt/convert/gather_nd_op.cc:16:
/usr/local/cuda/targets/x86_64-linux/include/cub/detail/device_synchronize.cuh:33: error: ignoring #pragma nv_exec_check_disable  [-Werror=unknown-pragmas]
 #pragma nv_exec_check_disable
 
In file included from /usr/local/cuda/targets/x86_64-linux/include/thrust/system/cuda/detail/util.h:36,
                 from /usr/local/cuda/targets/x86_64-linux/include/thrust/system/cuda/detail/internal/copy_cross_system.h:41,
                 from /usr/local/cuda/targets/x86_64-linux/include/thrust/system/cuda/detail/copy.h:100,
                 from /usr/local/cuda/targets/x86_64-linux/include/thrust/system/detail/adl/copy.h:42,
                 from /usr/local/cuda/targets/x86_64-linux/include/thrust/detail/copy.inl:22,
                 from /usr/local/cuda/targets/x86_64-linux/include/thrust/detail/copy.h:90,
                 from /usr/local/cuda/targets/x86_64-linux/include/thrust/detail/allocator/copy_construct_range.inl:21,
                 from /usr/local/cuda/targets/x86_64-linux/include/thrust/detail/allocator/copy_construct_range.h:45,
                 from /usr/local/cuda/targets/x86_64-linux/include/thrust/detail/contiguous_storage.inl:23,
                 from /usr/local/cuda/targets/x86_64-linux/include/thrust/detail/contiguous_storage.h:234,
                 from /usr/local/cuda/targets/x86_64-linux/include/thrust/detail/vector_base.h:30,
                 from /usr/local/cuda/targets/x86_64-linux/include/thrust/device_vector.h:26,
                 from /workspace/Paddle/paddle/fluid/inference/tensorrt/plugin/split_op_plugin.h:17,
                 from /workspace/Paddle/paddle/fluid/inference/tensorrt/convert/split_op.cc:16:
/usr/local/cuda/targets/x86_64-linux/include/cub/detail/device_synchronize.cuh:33: error: ignoring #pragma nv_exec_check_disable  [-Werror=unknown-pragmas]
 #pragma nv_exec_check_disable
```

Environment:
```
CUDA 11.6

```

This is the compilation command:
```
cmake .. -DWITH_PYTHON=OFF \
             -DWITH_GPU=ON \
             -DWITH_TESTING=OFF \
             -DWITH_INFERENCE_API_TEST=OFF \
             -DCMAKE_BUILD_TYPE=Release \
             -DCUDA_ARCH_NAME=Auto \
             -DON_INFER=ON \
             -DWITH_MKL=OFF \
             -DWITH_TENSORRT=ON \
             -DCMAKE_C_COMPILER=`which gcc-8` -DCMAKE_CXX_COMPILER=`which g++-8` \ 
&& make TARGET=SKYLAKEX -j`nproc`
```

@jiweibo

**Comment by jiweibo** [comment]
Add -DTENSORRT_ROOT=${your root of trt dir} to the cmake option and try.

Paddle should not have been verified on cuda11.6, currently only the version before cuda11.4 has been tested, maybe you can try to lower the cuda version

**Comment by jeng1220** [comment]
@erikguo ,
This is a known issue. The root-cause is CUB library in CUDA 11.6. There are 2 solutions to fix the error.
1. update CUB version to commit - [7d608bf](https://github.com/NVIDIA/cub/commit/7d608bf1dc14553e2fb219eabeed80b76621b6fe) in your local environment
2. cherry-pick PaddlePaddle's commit - [40694](https://github.com/PaddlePaddle/Paddle/pull/40694) to your local PaddlePaddle

**Comment by erikguo** [comment]
@jeng1220 
Thank you for your great answers, I chose the first choice to do the compilation. It works.

**Comment by erikguo** [comment]
> Add -DTENSORRT_ROOT=${your root of trt dir} to the cmake option and try.
> 
> Paddle should not have been verified on cuda11.6, currently only the version before cuda11.4 has been tested, maybe you can try to lower the cuda version

Thank you for help, after I added TENSORRT_ROOT, I can compile successfully.  However, when I tried to test with Resnet model example, I got  the following error.  Could you please give me some suggestions? thanx  @jiweibo  

I compiled paddle in the docker nvcr.io/nvidia/tritonserver:22.02-py3  (in which, CUDA 11.6 AND TENSORRT 8.2.3 :

I0404 13:46:31.718532 1 model_repository_manager.cc:994] loading: ResNet50-v1.5:1
I0404 13:46:31.819061 1 paddle.cc:1095] TRITONBACKEND_ModelInitialize: ResNet50-v1.5 (version 1)
I0404 13:46:31.825222 1 paddle.cc:1138] TRITONBACKEND_ModelInstanceInitialize: ResNet50-v1.5_0 (GPU device 0)
I0404 21:46:31.849813  3236 analysis_predictor.cc:849] TensorRT subgraph engine is enabled
--- Running analysis [ir_graph_build_pass]
--- Running analysis [ir_graph_clean_pass]
--- Running analysis [ir_analysis_pass]
terminate called after throwing an instance of 'phi::enforce::EnforceNotMet'
  what():  

--------------------------------------
C++ Traceback (most recent call last):
--------------------------------------
0   TRITONBACKEND_ModelInstanceInitialize
1   paddle_infer::CreatePredictor(paddle::AnalysisConfig const&)
2   paddle_infer::Predictor::Predictor(paddle::AnalysisConfig const&)
3   std::unique_ptr<paddle::PaddlePredictor, std::default_delete<paddle::PaddlePredictor> > paddle::CreatePaddlePredictor<paddle::AnalysisConfig, (paddle::PaddleEngineKind)2>(paddle::AnalysisConfig const&)
4   paddle::AnalysisPredictor::Init(std::shared_ptr<paddle::framework::Scope> const&, std::shared_ptr<paddle::framework::ProgramDesc> const&)
5   paddle::AnalysisPredictor::PrepareProgram(std::shared_ptr<paddle::framework::ProgramDesc> const&)
6   paddle::AnalysisPredictor::OptimizeInferenceProgram()
7   paddle::inference::analysis::Analyzer::RunAnalysis(paddle::inference::analysis::Argument*)
8   paddle::inference::analysis::IrAnalysisPass::RunImpl(paddle::inference::analysis::Argument*)
9   paddle::inference::analysis::IRPassManager::IRPassManager(paddle::inference::analysis::Argument*)
10  paddle::inference::analysis::IRPassManager::CreatePasses(paddle::inference::analysis::Argument*, std::vector<std::string, std::allocator<std::string > > const&)
11  paddle::framework::ir::PassRegistry::Get(std::string const&) const

----------------------
Error Message Summary:
----------------------
InvalidArgumentError: Pass tensorrt_subgraph_pass has not been registered. Please use the paddle inference library compiled with tensorrt or disable the tensorrt engine in inference configuration! 
  [Hint: Expected Has(pass_type) == true, but received Has(pass_type):0 != true:1.] (at /workspace/Paddle/paddle/fluid/framework/ir/pass.h:238)

Signal (6) received.
 0# 0x00005607981D7299 in tritonserver
 1# 0x00007F08ABA8D210 in /usr/lib/x86_64-linux-gnu/libc.so.6
 2# gsignal in /usr/lib/x86_64-linux-gnu/libc.so.6
 3# abort in /usr/lib/x86_64-linux-gnu/libc.so.6
 4# 0x00007F08ABE43911 in /usr/lib/x86_64-linux-gnu/libstdc++.so.6
 5# 0x00007F08ABE4F38C in /usr/lib/x86_64-linux-gnu/libstdc++.so.6
 6# 0x00007F08ABE4F3F7 in /usr/lib/x86_64-linux-gnu/libstdc++.so.6
 7# 0x00007F08ABE4F6A9 in /usr/lib/x86_64-linux-gnu/libstdc++.so.6
 8# paddle::framework::ir::PassRegistry::Get(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const in /opt/paddle/lib/libpaddle_inference.so
 9# paddle::inference::analysis::IRPassManager::CreatePasses(paddle::inference::analysis::Argument*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&) in /opt/paddle/lib/libpaddle_inference.so
10# paddle::inference::analysis::IRPassManager::IRPassManager(paddle::inference::analysis::Argument*) in /opt/paddle/lib/libpaddle_inference.so
11# paddle::inference::analysis::IrAnalysisPass::RunImpl(paddle::inference::analysis::Argument*) in /opt/paddle/lib/libpaddle_inference.so
12# paddle::inference::analysis::Analyzer::RunAnalysis(paddle::inference::analysis::Argument*) in /opt/paddle/lib/libpaddle_inference.so
13# paddle::AnalysisPredictor::OptimizeInferenceProgram() in /opt/paddle/lib/libpaddle_inference.so
14# paddle::AnalysisPredictor::PrepareProgram(std::shared_ptr<paddle::framework::ProgramDesc> const&) in /opt/paddle/lib/libpaddle_inference.so
15# paddle::AnalysisPredictor::Init(std::shared_ptr<paddle::framework::Scope> const&, std::shared_ptr<paddle::framework::ProgramDesc> const&) in /opt/paddle/lib/libpaddle_inference.so
16# std::unique_ptr<paddle::PaddlePredictor, std::default_delete<paddle::PaddlePredictor> > paddle::CreatePaddlePredictor<paddle::AnalysisConfig, (paddle::PaddleEngineKind)2>(paddle::AnalysisConfig const&) in /opt/paddle/lib/libpaddle_inference.so
17# paddle_infer::Predictor::Predictor(paddle::AnalysisConfig const&) in /opt/paddle/lib/libpaddle_inference.so
18# paddle_infer::CreatePredictor(paddle::AnalysisConfig const&) in /opt/paddle/lib/libpaddle_inference.so
19# 0x00007F07641015BC in /opt/tritonserver/backends/paddle/libtriton_paddle.so
20# 0x00007F0764102416 in /opt/tritonserver/backends/paddle/libtriton_paddle.so
21# 0x00007F0764105D36 in /opt/tritonserver/backends/paddle/libtriton_paddle.so
22# 0x00007F07641051EF in /opt/tritonserver/backends/paddle/libtriton_paddle.so
23# TRITONBACKEND_ModelInstanceInitialize in /opt/tritonserver/backends/paddle/libtriton_paddle.so
24# 0x00007F08AC62DDEE in /opt/tritonserver/bin/../lib/libtritonserver.so
25# 0x00007F08AC62F3B3 in /opt/tritonserver/bin/../lib/libtritonserver.so
26# 0x00007F08AC627067 in /opt/tritonserver/bin/../lib/libtritonserver.so
27# 0x00007F08AC4B07CA in /opt/tritonserver/bin/../lib/libtritonserver.so
28# 0x00007F08AC4BD9B1 in /opt/tritonserver/bin/../lib/libtritonserver.so
29# 0x00007F08ABE7BDE4 in /usr/lib/x86_64-linux-gnu/libstdc++.so.6
30# 0x00007F08AC2F9609 in /usr/lib/x86_64-linux-gnu/libpthread.so.0
31# clone in /usr/lib/x86_64-linux-gnu/libc.so.6

**Comment by erikguo** [comment]
I tried several methods, but all were failed.

First, I indicated the tensorrt root with -DTENSORRT_ROOT , it get the same error as above log.

Second, I tried to add protection in  "/workspace/Paddle/paddle/fluid/inference/tensorrt/plugin/gather_nd_op_plugin.h"  and "/workspace/Paddle/paddle/fluid/inference/tensorrt/plugin/split_op_plugin.h" as @jeng1220  recommended (cherry-pick PaddlePaddle's commit), it got the following new errors:

[ 54%] Building CUDA object paddle/fluid/inference/tensorrt/plugin/CMakeFiles/tensorrt_plugin.dir/split_op_plugin.cu.o
[ 54%] Building CXX object paddle/fluid/framework/CMakeFiles/executor_cache.dir/executor_cache.cc.o
[ 54%] Building CUDA object paddle/fluid/inference/tensorrt/plugin/CMakeFiles/tensorrt_plugin.dir/gather_nd_op_plugin.cu.o
/workspace/backend/Paddle/paddle/fluid/inference/tensorrt/plugin/split_op_plugin.h(96): error: qualified name is not allowed

/workspace/backend/Paddle/paddle/fluid/inference/tensorrt/plugin/split_op_plugin.h(96): error: this declaration has no storage class or type specifier

/workspace/backend/Paddle/paddle/fluid/inference/tensorrt/plugin/split_op_plugin.h(96): error: expected a ";"

/workspace/backend/Paddle/paddle/fluid/inference/tensorrt/plugin/split_op_plugin.h(97): error: qualified name is not allowed

/workspace/backend/Paddle/paddle/fluid/inference/tensorrt/plugin/split_op_plugin.h(97): error: this declaration has no storage class or type specifier

/workspace/backend/Paddle/paddle/fluid/inference/tensorrt/plugin/split_op_plugin.h(97): error: expected a ";"

/workspace/backend/Paddle/paddle/fluid/inference/tensorrt/plugin/split_op_plugin.h(29): error: exception specification for implicitly declared virtual function "paddle::inference::tensorrt::plugin::SplitPlugin::~SplitPlugin" is incompatible with that of overridden function "paddle::inference::tensorrt::plugin::PluginTensorRTV2Ext::~PluginTensorRTV2Ext"
          detected during:
            instantiation of "std::_Vector_base<_Tp, _Alloc>::_Vector_impl::_Vector_impl() [with _Tp=int, _Alloc=std::allocator<int>]" 
(31): here
            instantiation of class "std::_Vector_base<_Tp, _Alloc> [with _Tp=int, _Alloc=std::allocator<int>]" 
(31): here
            instantiation of class "std::vector<_Tp, _Alloc> [with _Tp=int, _Alloc=std::allocator<int>]" 
(31): here

/workspace/backend/Paddle/paddle/fluid/inference/tensorrt/plugin/split_op_plugin.h(29): error: exception specification for implicitly declared virtual function "paddle::inference::tensorrt::plugin::SplitPlugin::~SplitPlugin" is incompatible with that of overridden function "nvinfer1::IPluginV2Ext::~IPluginV2Ext"
          detected during:
            instantiation of "std::_Vector_base<_Tp, _Alloc>::_Vector_impl::_Vector_impl() [with _Tp=int, _Alloc=std::allocator<int>]" 
(31): here
            instantiation of class "std::_Vector_base<_Tp, _Alloc> [with _Tp=int, _Alloc=std::allocator<int>]" 
(31): here
            instantiation of class "std::vector<_Tp, _Alloc> [with _Tp=int, _Alloc=std::allocator<int>]" 
(31): here

/workspace/backend/Paddle/paddle/fluid/inference/tensorrt/plugin/split_op_plugin.h(29): error: exception specification for implicitly declared virtual function "paddle::inference::tensorrt::plugin::SplitPlugin::~SplitPlugin" is incompatible with that of overridden function "nvinfer1::IPluginV2::~IPluginV2"
          detected during:
            instantiation of "std::_Vector_base<_Tp, _Alloc>::_Vector_impl::_Vector_impl() [with _Tp=int, _Alloc=std::allocator<int>]" 
(31): here
            instantiation of class "std::_Vector_base<_Tp, _Alloc> [with _Tp=int, _Alloc=std::allocator<int>]" 
(31): here
            instantiation of class "std::vector<_Tp, _Alloc> [with _Tp=int, _Alloc=std::allocator<int>]" 
(31): here

/workspace/backend/Paddle/paddle/fluid/inference/tensorrt/plugin/split_op_plugin.cu(64): error: identifier "d_segment_offsets_" is undefined

/workspace/backend/Paddle/paddle/fluid/inference/tensorrt/plugin/split_op_plugin.cu(64): error: class "paddle::inference::tensorrt::plugin::SplitPlugin" has no member "d_segment_offsets_"

/workspace/backend/Paddle/paddle/fluid/inference/tensorrt/plugin/split_op_plugin.cu(66): error: identifier "d_output_ptrs_" is undefined

/workspace/backend/Paddle/paddle/fluid/inference/tensorrt/plugin/split_op_plugin.cu(66): error: class "paddle::inference::tensorrt::plugin::SplitPlugin" has no member "d_output_ptrs_"

/workspace/backend/Paddle/paddle/fluid/inference/tensorrt/plugin/split_op_plugin.cu(94): error: identifier "d_segment_offsets_" is undefined

/workspace/backend/Paddle/paddle/fluid/inference/tensorrt/plugin/split_op_plugin.cu(96): error: identifier "d_output_ptrs_" is undefined

/workspace/backend/Paddle/paddle/fluid/inference/tensorrt/plugin/split_op_plugin.cu(135): error: identifier "d_segment_offsets_" is undefined

/workspace/backend/Paddle/paddle/fluid/inference/tensorrt/plugin/split_op_plugin.cu(138): error: identifier "d_output_ptrs_" is undefined

/workspace/backend/Paddle/paddle/fluid/inference/tensorrt/plugin/split_op_plugin.cu(139): error: class "paddle::platform::details::ExternalApiType<<error-type>>" has no member "kSuccess"

/workspace/backend/Paddle/paddle/fluid/inference/tensorrt/plugin/split_op_plugin.cu(247): error: namespace "thrust" has no member "device_vector"

/workspace/backend/Paddle/paddle/fluid/inference/tensorrt/plugin/split_op_plugin.cu(247): error: type name is not allowed

/workspace/backend/Paddle/paddle/fluid/inference/tensorrt/plugin/split_op_plugin.cu(247): error: identifier "d_segment_offsets" is undefined

/workspace/backend/Paddle/paddle/fluid/inference/tensorrt/plugin/split_op_plugin.cu(259): error: namespace "thrust" has no member "device_vector"

/workspace/backend/Paddle/paddle/fluid/inference/tensorrt/plugin/split_op_plugin.cu(259): error: type name is not allowed

/workspace/backend/Paddle/paddle/fluid/inference/tensorrt/plugin/split_op_plugin.cu(259): error: expected an expression

/workspace/backend/Paddle/paddle/fluid/inference/tensorrt/plugin/split_op_plugin.cu(260): error: identifier "d_output_ptrs" is undefined

/workspace/backend/Paddle/paddle/fluid/inference/tensorrt/plugin/split_op_plugin.cu(266): error: class "paddle::platform::details::ExternalApiType<<error-type>>" has no member "kSuccess"

/workspace/backend/Paddle/paddle/fluid/inference/tensorrt/plugin/split_op_plugin.cu(275): error: namespace "thrust" has no member "device_vector"

/workspace/backend/Paddle/paddle/fluid/inference/tensorrt/plugin/split_op_plugin.cu(275): error: type name is not allowed

/workspace/backend/Paddle/paddle/fluid/inference/tensorrt/plugin/split_op_plugin.cu(275): error: expected an expression

/workspace/backend/Paddle/paddle/fluid/inference/tensorrt/plugin/split_op_plugin.cu(276): error: identifier "d_output_ptrs" is undefined

/workspace/backend/Paddle/paddle/fluid/inference/tensorrt/plugin/split_op_plugin.cu(282): error: class "paddle::platform::details::ExternalApiType<<error-type>>" has no member "kSuccess"

31 errors detected in the compilation of "/workspace/backend/Paddle/paddle/fluid/inference/tensorrt/plugin/split_op_plugin.cu".
make[2]: *** [paddle/fluid/inference/tensorrt/plugin/CMakeFiles/tensorrt_plugin.dir/build.make:90: paddle/fluid/inference/tensorrt/plugin/CMakeFiles/tensorrt_plugin.dir/split_op_plugin.cu.o] Error 1
make[2]: *** Waiting for unfinished jobs....


Third, I tried to update CUB version to commit  as @jeng1220 recommended, it also failed as the above new errors. 

So all routes is broken. Do you have some more suggestions ? Thank you very much   @jiweibo @jeng1220 

By the way, I'm tring to build paddle-backend with triton 22.02 version.

**Comment by heliqi** [comment]
Paddle has released CUDA 11.6 packs, please use paddle 2.4.
